### PR TITLE
Pattern bind reuse

### LINKED ISSUE #270: Smart matching doesn't influence type of variables
This should work, erlang makes sure both are the same integer value in this situation. 

```erlang
-module(test).

-spec test(integer() | undefined, integer()) -> integer().
test(I, I) -> I;
test(undefined, I) -> I;
test(I, _) -> I.
```

```
The variable on line 4 at column 9 is expected to have type integer() but it has type integer() | undefined

-spec test(integer() | undefined, integer()) -> integer().
test(I, I) -> I;
        ^
```

### PR DESCRIPTION
**Pull request by zuiderkwast**
Fixes #270 in the way outlined in #274.

TL;DR: Look at the test cases.

The capture/bind vars logic is revamped again. It is moved to an earlier stage, where all variables are extracted from the list of patterns. Therefore, Caps doesn't need to be passed around in add_type_pat/4 anymore.

* If a free variable is only occuring once in the pattern, it is handled as a free variable, i.e. it is not included in VEnv. (For bind_vars, any such variable is deleted from VEnv, to make sure it's treated as free later.)
* If a free variable occurs more than once in a pattern, it is bound to top(), which is later refined in add_type_pat/4 and add_any_types_pat/2. (For bind_vars, such variables are re-assigned to top() if they are bound previously, which is later refined in add_type_pat/4.)

### CODE REVIEW

**Comment by zuiderkwast** [comment]
Build fails because there is a crash in `make gradualize`. I'll look into it.

File: test/should_pass/intersection_with_any.erl [hunk 1]
```diff
@@ -10,3 +11,16 @@ any_refined_using_guard(X) when is_integer(X) -> X.
...
+-spec guess_two_dice(integer(), integer()) -> 0..6.
+guess_two_dice(X, Y) ->
+    case roll_two_dice() of
+        {X, Y} ->
+            X;
+        _GuessFail ->
+            0
+    end.
+
+%% untyped heper returning {1..6, 1..6}
```
Comment (josefs) [inline test/should_pass/intersection_with_any.erl]: "helper"

File: test/known_problems/should_fail/intersection_with_any_fail.erl [hunk 2]
```diff
@@ -14,3 +16,26 @@ any_refined_using_guard(X) when is_atom(X) ->
...
+        {Atom} ->
+            %% at this point Atom :: any()
+            %% but we want Atom :: atom() & any()
+            Atom
+    end.
+
+%% helper
+-spec get_any() -> any().
+get_any() ->
+    receive Any -> Any end.
```
Comment (josefs) [inline test/known_problems/should_fail/intersection_with_any_fail.erl]: Nice tests!

File: src/typechecker.erl [hunk 3]
```diff
@@ -3499,52 +3499,69 @@ type_check_function(Env, {function,_, Name, NArgs, Clauses}) ->
...
 %% by Pat. UBound is Ty or a subtype such that Pat :: UBound.
 add_types_pats(Pats, Tys, TEnv, VEnv, Caps) ->
-    add_types_pats(Pats, Tys, TEnv, VEnv, Caps, [], [], []).
-
-add_types_pats([], [], _TEnv, VEnv, _Caps, PatTysAcc, UBoundsAcc, CsAcc) ->
+    NewVEnv = assign_types_to_vars_bound_more_than_once(Pats, VEnv, Caps),
+    do_add_types_pats(Pats, Tys, TEnv, NewVEnv).
+
+%% Used in recursive calls from add_type_pat/4
+-spec do_add_types_pats(Pats :: [gradualizer_type:abstract_pattern()],
```
Comment (josefs) [inline src/typechecker.erl]: As I understand it, this function should not be called directly, all code should go via `add_types_pats` right? Can I suggest you put a comment on `add_types_pats` to highlight this fact

File: src/typechecker.erl [hunk 4]
```diff
@@ -3499,52 +3499,69 @@ type_check_function(Env, {function,_, Name, NArgs, Clauses}) ->
...
-		false ->
-		    throw({type_error, Var, VarTy, Ty})
+	#{A := VarTy} ->
+	    case glb(VarTy, Ty, TEnv) of
+                {?type(none), _Cs} ->
+                    %% TODO: Better type error (it's a pattern, not an
+                    %% expression)
+		    throw({type_error, Var, VarTy, Ty});
+		{RefinedTy, Cs} ->
+		    {type(none), RefinedTy, VEnv#{A := RefinedTy}, Cs}
```
Comment (josefs) [inline src/typechecker.erl]: It's nice to be able to get rid of the `capture_vars` logic here.

File: src/typechecker.erl [hunk 5]
```diff
@@ -3863,6 +3880,51 @@ add_any_types_pat({op, _, _Op, _Pat}, VEnv) ->
...
     VEnv.
 
+-spec assign_types_to_vars_bound_more_than_once(Pats :: [gradualizer_type:abstract_pattern()],
+                                                VEnv :: map(),
+                                                Caps :: capture_vars | bind_vars) ->
+                                                       NewVEnv :: map().
+%% Assigns the type top() to variables occurring more than once in the list of
+%% patterns. This prevents the variables from being treated as match-all, thus
+%% affecting refinement.
+assign_types_to_vars_bound_more_than_once(Pats, VEnv, Caps) ->
```
Comment (josefs) [inline src/typechecker.erl]: Giving variables the type `top()` feels a bit like a sledgehammer. I have the sense that we should be able to give them better types and refine the match-all logic to treat these variables better. But I don't have a concrete idea of how to do that.

File: test/known_problems/should_fail/intersection_with_any_fail.erl (hunk 2 above)
Comment (zuiderkwast) [reply test/known_problems/should_fail/intersection_with_any_fail.erl]: I just moved them from another test file. :-) They were failing as they should before but now they are known problems.

File: src/typechecker.erl (hunk 4 above)
Comment (zuiderkwast) [reply src/typechecker.erl]: :+1:

File: src/gradualizer_lib.erl [hunk 6]
```diff
@@ -119,3 +119,37 @@ pick_value(?type(range, [{_TagLo, _, neg_inf}, {_TagHi, _, Hi}])) ->
...
+fold_ast(_Fun, Acc, _LiteralEtc) ->
+    Acc.
+
+%% Returns the children of an AST node
+get_ast_children({clauses, Clauses}) ->
+    %% This one doesn't have an annotation
+    Clauses;
+get_ast_children(Node) ->
+    [_Tag, _Anno | Children] = tuple_to_list(Node),
+    Children.
```
[the remark this answers is not in the document]
Comment (zuiderkwast) [reply src/gradualizer_lib.erl]: You mean `{clauses, _}` is sneaky, right? Not that I sneakily copied this function from gradualizer_highlight? ;-) It should probably only be defined in one place....

File: src/typechecker.erl (hunk 3 above)
Comment (zuiderkwast) [reply src/typechecker.erl]: Good idea.

Btw, do we put the description above or below the spec?

File: src/typechecker.erl (hunk 5 above)
Comment (zuiderkwast) [reply src/typechecker.erl]: We known that these variables occurs in the pattern so we know tha they will be refined using `glb(Ty, top())`, which will be Ty in all cases. That's why top() is chosen here.

File: test/should_pass/intersection_with_any.erl (hunk 1 above)
Comment (zuiderkwast) [reply test/should_pass/intersection_with_any.erl]: Done

**Comment by FrankBro** [comment]
Wow that was fast. Awesome 🙌
